### ISSUE TITLE
Extremely high memory usage with deep images

### ISSUE THREAD

**Issue by jdiversen** [OpenImageIO/oiio#1602]
Is there a way to tweak the memory performance for deep operations --ch and --chnames?

We have been seeing spikes into the 80GB range processing deep images of 900MB on disk - which have multiple channels we're stripping and renaming.

Could you give us any advice on how to handle them? We have been running 1.7.7 and recently have tried 1.8.1 to no avail.

Example:
------------
```

/oiiotool --runstats deep_image.1291.exr --ch R,G,B,A,Z,ZBack,RenderID,multimatteID,objectId --chnames R,G,B,A,Z,ZBack,ID.RenderID,ID.multimatteID,ID.objectId -o test.exr

Threads: 24
oiiotool runtime statistics:
  Total time: 48.85s
      --ch         : 45.31
      --chnames    :  0.00
      -o           :  3.52
      input        :  0.01
      unaccounted  :  0.01
  Peak memory:    427.0 MB
  Current memory: 427.0 MB

OpenImageIO ImageCache statistics (0x1243fc0) ver 1.7.7
  Options:  max_memory_MB=4096.0 max_open_files=100 autotile=4096
            autoscanline=1 automip=0 forcefloat=1 accept_untiled=1
            accept_unmipped=1 read_before_insert=0 deduplicate=1
            unassociatedalpha=0 failure_retries=0
  Images : 1 unique
    ImageInputs : 1 created, 1 current, 1 peak
    Total pixel data size of all images referenced : 1.4 GB
    Total actual file size of all images referenced : 878.7 MB
    Pixel data read : 0 B
    File I/O time : 0.0s
    File open time only : 0.0s
    Peak cache memory : 0 B
  Image file statistics:
        opens   tiles    MB read   --redundant--   I/O time  res              File
      1    1        0        0.0                       0.0s  2258x1272x129.f32  deep_image.1291.exr  UNTILED

  Tot:     1        0        0.0   (    0    0.0)      0.0s
  1 not tiled, 1 not MIP-mapped
  Broken or invalid files: 0
```

Source image
------
This is the horror:

```
 $ oiinfo -v deep_image.1291.exr
deep_image.1291.exr : 2258 x 1272, 129 channel, deep half/half/half/half/float/float/uint/float/float/uint/uint/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half/half openexr
    channel list: R (half), G (half), B (half), A (half), Z (float), ZBack (float), RenderID (uint), WEIGHT (float), ZDepth (float), multimatteID (uint), objectId (uint), reflGloss (half), ALPHA.red (half), ALPHA.green (half), ALPHA.blue (half), GI.red (half), GI.green (half), GI.blue (half), Pobject.red (half), Pobject.green (half), Pobject.blue (half), Pref.red (half), Pref.green (half), Pref.blue (half), Pworld.red (half), Pworld.green (half), Pworld.blue (half), SSS.red (half), SSS.green (half), SSS.blue (half), TRANSPARENCY.red (half), TRANSPARENCY.green (half), TRANSPARENCY.blue (half), diffuse.red (half), diffuse.green (half), diffuse.blue (half), lgt_dome.red (half), lgt_dome.green (half), lgt_dome.blue (half), lgt_key.red (half), lgt_key.green (half), lgt_key.blue (half), lighting.red (half), lighting.green (half), lighting.blue (half), reflect.red (half), reflect.green (half), reflect.blue (half), reflectionFilter.red (half), reflectionFilter.green (half), reflectionFilter.blue (half), refract.red (half), refract.green (half), refract.blue (half), sampleRate.red (half), sampleRate.green (half), sampleRate.blue (half), selfIllum.red (half), selfIllum.green (half), selfIllum.blue (half), specular.red (half), specular.green (half), specular.blue (half), tx_rgbMask01.red (half), tx_rgbMask01.green (half), tx_rgbMask01.blue (half), tx_rgbMask010.red (half), tx_rgbMask010.green (half), tx_rgbMask010.blue (half), tx_rgbMask011.red (half), tx_rgbMask011.green (half), tx_rgbMask011.blue (half), tx_rgbMask012.red (half), tx_rgbMask012.green (half), tx_rgbMask012.blue (half), tx_rgbMask013.red (half), tx_rgbMask013.green (half), tx_rgbMask013.blue (half), tx_rgbMask014.red (half), tx_rgbMask014.green (half), tx_rgbMask014.blue (half), tx_rgbMask015.red (half), tx_rgbMask015.green (half), tx_rgbMask015.blue (half), tx_rgbMask016.red (half), tx_rgbMask016.green (half), tx_rgbMask016.blue (half), tx_rgbMask02.red (half), tx_rgbMask02.green (half), tx_rgbMask02.blue (half), tx_rgbMask03.red (half), tx_rgbMask03.green (half), tx_rgbMask03.blue (half), tx_rgbMask04.red (half), tx_rgbMask04.green (half), tx_rgbMask04.blue (half), tx_rgbMask05.red (half), tx_rgbMask05.green (half), tx_rgbMask05.blue (half), tx_rgbMask06.red (half), tx_rgbMask06.green (half), tx_rgbMask06.blue (half), tx_rgbMask07.red (half), tx_rgbMask07.green (half), tx_rgbMask07.blue (half), tx_rgbMask08.red (half), tx_rgbMask08.green (half), tx_rgbMask08.blue (half), tx_rgbMask09.red (half), tx_rgbMask09.green (half), tx_rgbMask09.blue (half), tx_rgbMask17.red (half), tx_rgbMask17.green (half), tx_rgbMask17.blue (half), tx_rgbMask18.red (half), tx_rgbMask18.green (half), tx_rgbMask18.blue (half), tx_rgbMask19.red (half), tx_rgbMask19.green (half), tx_rgbMask19.blue (half), tx_rgbMask20.red (half), tx_rgbMask20.green (half), tx_rgbMask20.blue (half), uvcoords.red (half), uvcoords.green (half), uvcoords.blue (half), worldYcameraZ.red (half), worldYcameraZ.green (half), worldYcameraZ.blue (half)
    oiio:ColorSpace: "Linear"
    compression: "zips"
    cameraAperture: 29.5851
    cameraFNumber: 8
    cameraFarClip: 1e+30
    cameraFarRange: 1e+18
    cameraFocalLength: 20
    cameraFov: 72.9753
    cameraNearClip: 0
    cameraNearRange: 0
    cameraProjection: 0
    cameraTargetDistance: 200
    worldtocamera: -0.177762 7.45058e-09 0.984074 0 -0.981349 0.0743603 -0.17727 0 -0.073176 -0.997231 -0.0132184 0 -5580.17 33.3102 636.454 1
    openexr:chunkCount: 1272
    PixelAspectRatio: 1
    screenWindowCenter: 0 0
    screenWindowWidth: 1
    openexr:version: 1
```

Conclusion
------

This process bounces around 4Gb but spikes out at around 15Gb, watching with 'top'.  Our worst offender took 20mins to process off the network and 86GB of RAM.

If you have any advice for how we can modify how it executes / caches / etc, we'd love to try it.

Many thanks,
Jason

**Comment by lgritz** [comment]
I know this is old. Apologies.

The problem is that the way --ch works by default is to read in the whole file into memory (all channels), and then make a second image for the selected channels and copy them.

I think that there is a partial solution available. This seems to be a duplicate of #1537, which was addressed by PR #1541. If you use the explicit `-i` (input image) command, there is an optional modifier `ch=` that will restrict the channels read. So, for example,

    oiiotool --runstats  -i:ch=R,G,B,A,Z,ZBack,RenderID,multimatteID,objectId deep_image.1291.exr -o test.exr

I *think* that will do what you want, and should be many many times faster and lower memory, since you are asking for a tiny fraction of the total channels.

There are two caveats, however.

1. #1537/#1541 were about flat files, not deep. A quick glance at the code indicates that it *ought* to work for deep files as well, but I haven't tested with one of these monstrous files to be sure.

2. How well it performs will depend on how spread out in the file your desired channels are. YMMV.

Let me further explain point (2). The OIIO ImageInput::read_native_deep_stuff APIs don't take a list of arbitrary channels, they just take a numeric [chbegin, chend) range. So what `oiiotool -i:ch=...` does is find the minimum and maximum channel indices in your list and restrict the actual disk read and decode to that range (which could be as tight as consisting only of your desired channels, or if you are "skipping around a lot" it could read a bunch of channels you don't want), and then in a second step, if there are those "unwanted channels" in your range, it does the usual ImageBufAlgo::channels() to copy just the ones you need (much like a raw `oiiotool --ch` would). 

OpenEXR stores channels in alphabetical order (ugh!), but OIIO reshuffles so that R, G, B, A, Z, Zback are at front, then everything else follows, grouped by "layer name" if there is one, but otherwise in whatever (alphabetical?) order openexr used.  Now, it looks like in the example you posted above, the channels you wanted are all close to the front, so I would expect it to perform VERY well. But if in some other example you happen to want both R (the first channel) and worldYcameraZ.blue (the last channel), you won't get any better performance than the regular `--ch`, if you follow my logic.

**Comment by lgritz** [comment]
I'm going to close this issue, since I think that the solution I outlined will get you where you need to be (and for all I know, perhaps you discovered it on your own shortly after posting this, since you never followed up to poke me).

If this is still inadequate, please re-open or file a new issue (don't forget to point out that I "owe you one" for taking so long on this), and I'll see what we can do to be even more efficient about it.
